pydantic_ai_slim/pydantic_ai/providers/gateway.py

```diff
@@ -0,0 +1,125 @@
+"""This module implements the Pydantic AI Gateway provider."""
+
+from __future__ import annotations as _annotations
+
+import os
+from typing import TYPE_CHECKING, Any, Literal, overload
+
+import httpx
+
+from pydantic_ai.exceptions import UserError
+from pydantic_ai.models import get_user_agent
+from pydantic_ai.profiles import ModelProfile
+from pydantic_ai.providers import InterfaceClient, Provider
+
+if TYPE_CHECKING:
+    from google.auth.credentials import Credentials
+    from google.genai import Client as GoogleClient
+    from groq import AsyncGroq
+    from openai import AsyncOpenAI
+
+    from pydantic_ai.providers.google import VertexAILocation
+
+
+class GatewayProvider(Provider[InterfaceClient]):
+    """Provider to access the Pydantic AI Gateway."""
+
+    @property
+    def name(self) -> str:
+        # TODO(Marcelo): This is actually NEVER used, because the `provider` is never `GatewayProvider`, it's always
+        # the actual provider class: `OpenAIProvider`, `GroqProvider` and `GoogleProvider`. Maybe it's fine?
+        return 'gateway'
+
+    @property
+    def base_url(self) -> str: ...
+
+    @property
+    def client(self) -> InterfaceClient: ...
+
+    def model_profile(self, model_name: str) -> ModelProfile | None: ...
+
+    @overload
+    def __new__(
+        cls,
+        provider: Literal['openai'],
+        *,
+        api_key: str | None = None,
+        base_url: str | None = None,
+        http_client: httpx.AsyncClient | None = None,
+    ) -> Provider[AsyncOpenAI]: ...
+
+    @overload
+    def __new__(
+        cls,
+        provider: Literal['groq'],
+        *,
+        api_key: str | None = None,
+        base_url: str | None = None,
+        http_client: httpx.AsyncClient | None = None,
+    ) -> Provider[AsyncGroq]: ...
+
+    @overload
+    def __new__(
+        cls,
+        provider: Literal['google-vertex'],
+        *,
+        api_key: str | None = None,
+        base_url: str | None = None,
+        credentials: Credentials | None = None,
+        project: str | None = None,
+        location: VertexAILocation | Literal['global'] | None = None,
+    ) -> Provider[GoogleClient]: ...
+
+    def __new__(
+        cls,
+        provider: Literal['openai', 'groq', 'google-vertex'] | str,
```

Comment:
It doesn't look like this arg is being passed in anywhere if the model class just gets `provider='gateway'`?

Reply:
I've dropped the `'gateway'` string from models.